File: apps/data-api/src/aeromaint_api/domain/clock.py
from collections.abc import Sequence
from dataclasses import dataclass

from aeromaint_api.domain.manifest import ClockDefinition, StreamGap


@dataclass(frozen=True)
class IndexedFrame:
    frame_number: int
    presentation_ns: int
    keyframe: bool


def map_to_session_time(source_ns: int, clock: ClockDefinition) -> int:
    return (
        clock.session_epoch_ns
        + ((source_ns - clock.source_epoch_ns) * clock.rate_numerator) // clock.rate_denominator
    )


def _inside_gap(time_ns: int, gaps: Sequence[StreamGap]) -> bool:
    return any(gap.start_ns <= time_ns < gap.end_ns for gap in gaps)


def frame_at_or_before(
    frames: Sequence[IndexedFrame], requested_ns: int, gaps: Sequence[StreamGap] = ()
) -> IndexedFrame | None:
    if _inside_gap(requested_ns, gaps):
        return None
    match = None
    for frame in frames:
        if frame.presentation_ns > requested_ns:
            break
        match = frame
    return match


def nearest_frame(
    frames: Sequence[IndexedFrame], requested_ns: int, gaps: Sequence[StreamGap] = ()
) -> IndexedFrame | None:
    if _inside_gap(requested_ns, gaps):
        return None
    return min(
        frames,
        key=lambda frame: (abs(frame.presentation_ns - requested_ns), frame.presentation_ns),
        default=None,
    )

File: packages/contracts/src/clock.ts
import type { ClockDefinition, StreamGap, TimestampNs } from "./index.js";

export interface IndexedFrame {
  readonly frameNumber: number;
  readonly presentationNs: TimestampNs;
  readonly keyframe: boolean;
}

function floorDivide(numerator: bigint, denominator: bigint): bigint {
  const quotient = numerator / denominator;
  const remainder = numerator % denominator;
  return remainder !== 0n && numerator < 0n ? quotient - 1n : quotient;
}

export function mapToSessionTime(
  sourceNs: TimestampNs,
  clock: ClockDefinition
): TimestampNs {
  const scaled = floorDivide(
    (sourceNs - clock.sourceEpochNs) * BigInt(clock.rateNumerator),
    BigInt(clock.rateDenominator)
  );
  return clock.sessionEpochNs + scaled;
}

function isInsideGap(timeNs: TimestampNs, gaps: readonly StreamGap[]): boolean {
  return gaps.some((gap) => timeNs >= gap.startNs && timeNs < gap.endNs);
}

export function frameAtOrBefore(
  frames: readonly IndexedFrame[],
  requestedNs: TimestampNs,
  gaps: readonly StreamGap[] = []
): IndexedFrame | undefined {
  if (isInsideGap(requestedNs, gaps)) return undefined;
  let match: IndexedFrame | undefined;
  for (const frame of frames) {
    if (frame.presentationNs > requestedNs) break;
    match = frame;
  }
  return match;
}

export function nearestFrame(
  frames: readonly IndexedFrame[],
  requestedNs: TimestampNs,
  gaps: readonly StreamGap[] = []
): IndexedFrame | undefined {
  if (isInsideGap(requestedNs, gaps)) return undefined;
  let match: IndexedFrame | undefined;
  let distance: bigint | undefined;
  for (const frame of frames) {
    const candidateDistance =
      frame.presentationNs >= requestedNs
        ? frame.presentationNs - requestedNs
        : requestedNs - frame.presentationNs;
    if (distance === undefined || candidateDistance < distance) {
      match = frame;
      distance = candidateDistance;
    }
  }
  return match;
}
